plugins/PagePerformance/Metrics.php

```diff
@@ -0,0 +1,126 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Plugins\PagePerformance;
+
+use Piwik\Plugin\Dimension\ActionDimension;
+use Piwik\Plugins\PagePerformance\Columns\Metrics\AveragePageLoadTime;
+use Piwik\Plugins\PagePerformance\Columns\Metrics\AverageTimeDomCompletion;
+use Piwik\Plugins\PagePerformance\Columns\Metrics\AverageTimeDomProcessing;
+use Piwik\Plugins\PagePerformance\Columns\Metrics\AverageTimeLatency;
+use Piwik\Plugins\PagePerformance\Columns\Metrics\AverageTimeOnLoad;
+use Piwik\Plugins\PagePerformance\Columns\Metrics\AverageTimeTransfer;
+use Piwik\Plugins\PagePerformance\Columns\TimeDomCompletion;
+use Piwik\Plugins\PagePerformance\Columns\TimeDomProcessing;
+use Piwik\Plugins\PagePerformance\Columns\TimeLatency;
+use Piwik\Plugins\PagePerformance\Columns\TimeOnLoad;
+use Piwik\Plugins\PagePerformance\Columns\TimeTransfer;
+
+class Metrics
+{
+    /**
+     * @return \Piwik\Plugins\PagePerformance\Columns\Metrics\AveragePerformanceMetric[]
+     */
+    public static function getPagePerformanceMetrics()
+    {
+        $metrics = [
+            new AverageTimeLatency(),
+            new AverageTimeTransfer(),
+            new AverageTimeDomProcessing(),
+            new AverageTimeDomCompletion(),
+            new AverageTimeOnLoad(),
+        ];
+
+        $mappedMetrics = [];
+
+        foreach ($metrics as $metric) {
+            $mappedMetrics[$metric->getName()] = $metric;
+        }
+
+        return $mappedMetrics;
+    }
+
+    /**
+     * @return \Piwik\Plugins\PagePerformance\Columns\Metrics\AveragePerformanceMetric[]
+     */
+    public static function getAllPagePerformanceMetrics()
+    {
+        $metrics = [
+            new AverageTimeLatency(),
```

Comment:
We should also show the Avg Generation Time here? As mentioned somewhere else a `server time` be great as well if that's not there yet (if it's there already, maybe the naming could be better?) 

`Avg Generation Time` I also always found very confusing the naming. I would maybe think this would be either server time or the time it takes to fully render the page including server time. But not what it currently is... wonder if there's a better name.

Reply:
The performance timings basically has this metrics:

![](https://www.tripadvisor.com/engineering/wp-content/uploads/2015/08/timing-overview.png)

The current generation time is the time from `requestStart` to `responseEnd`. So it's basically the time from the browser sending the request, to the browser having received the complete answer. To be honest - I think that metric is not very useful as you imho can't work with it in any way. You don't know if the server needs long or if the network was slow. I would actually even suggest to remove it

The time the server needs to process the request is currently included in `Latency`, together with DNS and TCP overhead, which should be very low in most cases. If you think that would be useful, we could split up the Latency time into `Network Latency` and `Page Generation Time`. Even though that would mean the `Network Latency` would be actually `0` quite often, as DNS/TCP gets cached after the first request.

The metrics we currently have can be simply summed up to get the full time it took to render the page completely. That's the reason why I did not include the current page generation time anywhere in the new reports.

Reply:
@sgiehl agreed. I've always found this metric useless too and not helping. I wonder if there's an alternative to remove it since some users would maybe want it simply for historical reasons. Not sure how to best do that though. Like mentioning in the description this metric is deprecated and only there for historical reasons. We could include one useful performance metric in the regular table report, then have this metric also in the performance report but could list it last or so. And for new installs we could completely remove it based on `DbHelper::getInstallVersion()`. Just a thought

Reply:
I'm not sure how to handle that best either. Have added the page generation time to the page performance table as last column if Matomo was installed before 4.0. If Matomo was installed later the page generation time won't be shown there and will now also be hidden in the page url / title reports and the visit summary sparkline for it.
Have not yet added some kind of note that the page generation time metric is deprecated. Currently we would still go on with tracking it. Wondering if we should remove that in that case 🤔

Reply:
You mean no longer track it if it was installed from Matomo 4? That could work if it's easy to do. Would need to make sure this option is cached in tracker. Eventually it could allow using the custom_float for something else I suppose. We could maybe also just completely stop tracking it maybe. Considering they could still get kind of the same value (by aggregating the new metrics) and compare it to previous value manually... 
They would keep the historical data also... it may confuse people though when it now always says 0 for average generation time once they upgrade. Could the avg generation time for these people be a computed/processed metric aggregating two metrics like server time = transfer time? Not sure if that be easy to do or make sense...

Reply:
Ok. To be able to calculate the current page generation time with the performance metrics, we would need to split up latency into `network` and `server` latency. Page generation could then be calculated with `server latency + transfer time`. But actually I don't see any benefits in being able to compute a metric that is still kind of useless. We could also simply sum up latency and transfer time, but that would mean also network latency is included, resulting in a bit higher numbers. If someone really tries to use those number, having higher number after the update of Matomo might be bit confusing...

Reply:
Splitting up latency into network and server might be beneficial independent of this? AKA `server` time = `'(performanceAlias.timing.responseStart - performanceAlias.timing.requestStart);` and network time = `'(performanceAlias.timing.requestStart - performanceAlias.timing.fetchStart);` ?

I guess typically the network time be quite low. Like `2ms` or something. But it could also be higher eg when having users around the world and no CDN, etc and would just help better understand where issues might be?

Calculating that metric indeed wouldn't be needed I'd say. I'd say we could also stop tracking that new metric and if people are still interested in it, then they could be looking at the other metrics to compare things. 

The problem be still that users who installed Matomo before 4.0 would see `0` in some newer reports. 
* We could set a date when they upgraded to Matomo 4 in the options table and if a start date of a report is past that day, remove that metric. 
* Or we could iterate over all columns and should all values be `0` then we remove the column. However this might be a bit slow?

Reply:
> I guess typically the network time be quite low. Like 2ms or something. But it could also be higher eg when having users around the world and no CDN, etc and would just help better understand where issues might be?

Correct. It should be very low or even zero for most requests, as after the first request DNS/TCP is cached and won't take any time. Can split up that metric if you prefer, but that would mean another couple of hours of work to adjust everything. Let me know, and I'll do it.

> I'd say we could also stop tracking that new metric

Will remove that everywhere. And add a deprecation note

> Or we could iterate over all columns and should all values be 0 then we remove the column. However this might be a bit slow?

That's kind of already done before computing the metric:
https://github.com/matomo-org/matomo/blob/b16a791aa3650d85af829156129c2bd44c7cb075/plugins/Actions/Columns/Metrics/AveragePageGenerationTime.php#L66-L101

But the columns aren't removed from being displayed, as it seems to be done too early. Seems that is broken for longer time already. Will move removing the columns from `columns_to_display` to a later point. Then it should hide the columns if they don't have a value in the report.

Reply:
> I'd say we could also stop tracking that new metric

Just realized, that it was also possible to manually set that metric, which might maybe be used by some people to track the time the server actually spent. Will remove the part, that calculates the metric based on browser timings. Or shall we also remove the possibility to set that manually?

Reply:
> Or shall we also remove the possibility to set that manually?

I reckon should someone set it, then we could log a warning that this feature is deprecated in favour of the new ones? Would need to get the new metrics also implemented ideally by other SDKs (android/ios/java/...). Some of the metrics may be useful for them eg the load time etc.

>  Can split up that metric if you prefer, but that would mean another couple of hours of work to adjust everything. Let me know, and I'll do it.

It be really great not spending more time on this but I think it's really valuable splitting it up simply because it helps a great deal when trying to act based on this data as network and server are 2 totally different problems and you wouldn't know where the actual problem was. Be good to add and hopefully it's quick to add.